crates/mcp-client/src/transport/sse.rs

```diff
@@ -0,0 +1,189 @@
+        while let Ok(Some(event)) = stream.try_next().await {
+            match event {
+                SSE::Event(event) if event.event_type == "endpoint" => {
+                    let base_url = sse_url.trim_end_matches('/').trim_end_matches("sse");
+                    let endpoint_path = event.data.trim_start_matches('/');
+                    let post_url = format!("{}{}", base_url, endpoint_path);
+                    println!("Endpoint for POST requests: {}", post_url);
+                    *post_endpoint.lock().await = Some(post_url);
+                    break;
+                }
+                _ => continue,
+            }
+        }
+
+        // Now handle all subsequent messages
+        let message_handler = tokio::spawn({
+            let pending_requests = pending_requests.clone();
+            async move {
+                while let Ok(Some(event)) = stream.try_next().await {
+                    match event {
+                        SSE::Event(event) if event.event_type == "message" => {
+                            if let Ok(message) = serde_json::from_str::<JsonRpcMessage>(&event.data)
+                            {
+                                Self::handle_message(message, pending_requests.clone()).await;
+                            }
+                        }
+                        _ => continue,
+                    }
+                }
+            }
+        });
+
+        // Process outgoing messages
+        while let Some(transport_msg) = message_rx.recv().await {
+            let post_url = match post_endpoint.lock().await.as_ref() {
+                Some(url) => url.clone(),
+                None => {
+                    eprintln!("No POST endpoint available");
+                    continue;
+                }
+            };
+
+            // Store response channel if this is a request
+            if let Some(response_tx) = transport_msg.response_tx {
+                if let JsonRpcMessage::Request(request) = &transport_msg.message {
+                    if let Some(id) = &request.id {
+                        pending_requests
+                            .lock()
+                            .await
+                            .insert(id.to_string(), response_tx);
+                    }
+                }
+            }
+
+            // Send message via HTTP POST
+            let message_str = match serde_json::to_string(&transport_msg.message) {
+                Ok(s) => s,
+                Err(e) => {
+                    eprintln!("Failed to serialize message: {}", e);
+                    continue;
+                }
+            };
+
+            if let Err(e) = http_client
+                .post(&post_url)
+                .header("Content-Type", "application/json")
+                .body(message_str)
+                .send()
+                .await
+            {
+                eprintln!("Failed to send message: {}", e);
+            }
+        }
```

Comment:
Should we push back an `Err` that encapsulates the http_client or serde error back through the channel to the user and ensure there's some level of error handling?

Looking at the [Kotlin SDK for SSE](https://github.com/modelcontextprotocol/kotlin-sdk/blob/main/src/main/kotlin/io/modelcontextprotocol/kotlin/sdk/client/SSEClientTransport.kt) seems like they call the `onError` handler or even `throw e` back up.